lib/documents.js

```diff
@@ -2263,12 +2264,143 @@ function writeAllDocumentsImpl(options) {
   let operation = new Operation(
       'read forestInfo', this.client, requestOptions, 'empty', 'single'
   );
-  operation.stream = inputStream;
-  let result = requester.startRequest(operation);
-  operation.response = result;
+
+  let jobState = {
+    docInstance : this,
+    stream: inputStream,
+    requesterCount: 0,
+    docsWrittenSuccessfully:0,
+    docsFailedToBeWritten: 0,
+    jobOptions: jobOptions
+  };
+
+  operation.jobState = jobState;
+
+  requester.startRequest(operation).result(mlutil.callbackOn(operation, onWriteAllInit))
+      .catch((err) => operation.stream.emit('error', err));
+
   return inputStream;
 }
 
+function finishWriter(operation) {
+  operation.jobState.requesterCount--;
+
+  if(operation.jobState.requesterCount<1) {
+
+    if(operation.jobState.jobOptions && operation.jobState.jobOptions.onCompletion){
+
+      let summary = ('Number of documents written successfully = '+
+          ((operation.jobState.docsWrittenSuccessfully>0)?operation.jobState.docsWrittenSuccessfully:0)+'\n'+
+          'Number of documents failed to be written = '+
+          ((operation.jobState.docsFailedToBeWritten>0)?operation.jobState.docsFailedToBeWritten:0)
+      );
+      if(operation.jobState.error){
+        summary = summary + 'Processing was cancelled by an error - '+operation.error.toString();
+      }
+      operation.jobState.jobOptions.onCompletion(summary);
+    }
+
+  }
+}
+
+function onWriteAllInit(output) {
+
+  let maxRequesters = output.length;
+  const operation = this;
+  const jobState = operation.jobState;
+
+  if(jobState.jobOptions && jobState.jobOptions.concurrentRequests) {
+    if(jobState.jobOptions.concurrentRequests.multipleOf &&
+        jobState.jobOptions.concurrentRequests.multipleOf === 'hosts'){
+      const hosts = new Set();
+      for(let i=0; i<output.length; i++) {
+        hosts.add(output[i].host);
+      }
+      maxRequesters = hosts.size;
+    }
+    const multiplier = (jobState.jobOptions.concurrentRequests.multiplier &&
+        jobState.jobOptions.concurrentRequests.multiplier>0)?
+        jobState.jobOptions.concurrentRequests.multiplier:0;
+    maxRequesters = (multiplier >0) ? (maxRequesters*multiplier):maxRequesters;
+
+  }
+
+  for (let i = 0; i < maxRequesters; i++) {
+    jobState.requesterCount++;
+    onWriteAllDocs(operation,i);
+  }
+
+}
+
+function onWriteAllDocs(operation, writerId) {
+
+  const writeBatchArray = [];
+  const jobState = operation.jobState;
+  const batchSize = (jobState.jobOptions && jobState.jobOptions.batchSize &&
+      jobState.jobOptions.batchSize >0) ?
+      jobState.jobOptions.batchSize:100;
+
+  a:
+      for (let i = 0; i < batchSize; i++) {
+        const record = jobState.stream.read();
+
+        if(jobState.jobOptions && jobState.jobOptions.defaultMetadata) {
+          writeBatchArray.push(jobState.jobOptions.defaultMetadata);
+        }
```

Comment:
Yes, that's the intent of jobOptions.defaultMetadata

It will be necessary to validate the properties of the defaultMetadata descriptor in writeAllDocumentsImpl().  The metadata descriptor can take the same properties as the document descriptor except uri and content.